### PR TITLE
TLS TrustStore with K8S cert

### PR DESCRIPTION
**Pull request by kousik93**
If found, add Kubernetes CA cert to TLS TrustStore. Else, use default. Ignore if TrustStore is provided in config file.

### CODE REVIEW

**Comment by kousik93** [comment]
@johngmyers Can you please review this pull request. Thank You.

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 1]
```diff
@@ -153,6 +164,62 @@ public static void main(String[] args)
...
+                certstream.close();
+
+                // Add the certificate
+                keystore.setCertificateEntry(alias, certs);
+
+                // Save the new truststore contents
+                FileOutputStream out = new FileOutputStream(destinationTrustStorePath);
+                keystore.store(out, password);
+                out.close();
+                System.out.println("Created new TrustStore with Kubernetes Certificate");
```
Comment (johngmyers) [inline launcher/src/main/java/com/proofpoint/launcher/Main.java]: The logging is a bit verbose. Perhaps one log line max, only if "verbose" is true?

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 2]
```diff
@@ -153,6 +164,62 @@ public static void main(String[] args)
...
+                return defaultTrustStorePath;
+            }
+        }
+
+        private InputStream getCertStream ( String fname ) throws IOException {
+            FileInputStream fis = new FileInputStream(fname);
+            DataInputStream dis = new DataInputStream(fis);
+            byte[] bytes = new byte[dis.available()];
+            dis.readFully(bytes);
+            ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
```
Comment (johngmyers) [inline launcher/src/main/java/com/proofpoint/launcher/Main.java]: Why are you buffering this?

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 3]
```diff
@@ -396,6 +463,25 @@ else if (line.matches("(?i)\\s*#\\s*allow\\s+spaces\\s*")) {
...
 
+            //Add the trust store
+            try {
+                Boolean addTrustStoreConfig = true;
+                //If the trustStore arg is passed from config file, ignore this new truststore.
+                for(String s : launcherArgs){
+                    if(s.contains("-Djavax.net.ssl.trustStore=")){
+                        addTrustStoreConfig = false;
+                    }
+                }
```
Comment (johngmyers) [inline launcher/src/main/java/com/proofpoint/launcher/Main.java]: `launcherArgs.stream().noneMatch(s -> s.startsWith("-Djavax.net.ssl.trustStore="))`

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 4]
```diff
@@ -396,6 +463,25 @@ else if (line.matches("(?i)\\s*#\\s*allow\\s+spaces\\s*")) {
...
+                if(addTrustStoreConfig){
+                    javaArgs.add("-Djavax.net.ssl.trustStore=" + addKubernetesToTrustStore());
+                }else{
+                    System.out.println("Using trustStore provided in config");
```
Comment (johngmyers) [inline launcher/src/main/java/com/proofpoint/launcher/Main.java]: No need to log.

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 5]
```diff
@@ -396,6 +463,25 @@ else if (line.matches("(?i)\\s*#\\s*allow\\s+spaces\\s*")) {
...
+                }
+            }catch(IOException | CertificateException | KeyStoreException | NoSuchAlgorithmException e){
+                System.err.println("Failed to get Truststore when adding Kubernetes cert. Error!");
```
Comment (johngmyers) [inline launcher/src/main/java/com/proofpoint/launcher/Main.java]: Doesn't log what the error was. "Error!" is redundant. Text is possibly misleading.

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 6]
```diff
@@ -153,6 +164,62 @@ public static void main(String[] args)
             }
         }
 
+        // Copies default store to accessable path. Adds k8s ca root cert to truststore. Returns path of new truststore
+        private String addKubernetesToTrustStore() throws IOException, FileNotFoundException, CertificateException, KeyStoreException, NoSuchAlgorithmException{
+            String javaHome = System.getenv("JAVA_HOME");
+            String defaultTrustStorePath = javaHome + "/jre/lib/security/cacerts";
```
[the remark this answers is not in the document]
Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: It is not necessarily possible to detect the current truststore path according to [This](https://stackoverflow.com/questions/13657299/how-to-find-current-truststore-on-disk-programatically). Java's default TrustStore is located at $JAVA_HOME/jre/lib/security 
If K8S cert is found, I'm thinking to error out and ask the user to specify JAVA_HOME

File: launcher/src/main/java/com/proofpoint/launcher/Main.java (hunk 2 above)
Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: Will change this to BufferedInputStream. This supports mark and reset options so that call to generateCertificate consumes only one certificate at a time.

**Comment by johngmyers** [comment]
Commit message is unnecessarily long and contains unnecessary details.

File: CHANGES [hunk 7]
```diff
@@ -13,6 +13,12 @@ Platform 1.67
   -XX:UseCGroupMemoryLimitForHeap argument to set the heap size limit to the
   Docker container's memory limit.
 
+  If running inside Kubernetes, the launcher automatically detect the CA cert
+  at path /var/run/secrets/kubernetes.io/serviceaccount/ca.crt and add to the default TLS
+  TrustStore. The location of new TrustStore with the K8S cert added is /service/var/cacerts
+  If the K8S cert is not present, default TrustStore will be used.
+  If TrustStore location is specified in config file, will use that and ignore others.
```
[the remark this answers is not in the document]
Comment (kousik93) [reply CHANGES]: Done

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 8]
```diff
@@ -153,6 +164,62 @@ public static void main(String[] args)
             }
         }
 
+        // Copies default store to accessable path. Adds k8s ca root cert to truststore. Returns path of new truststore
+        private String addKubernetesToTrustStore() throws IOException, FileNotFoundException, CertificateException, KeyStoreException, NoSuchAlgorithmException{
```
[the remark this answers is not in the document]
Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: Done. Wrapped exceptions with RuntimeException.

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 9]
```diff
@@ -153,6 +164,62 @@ public static void main(String[] args)
...
+            String javaHome = System.getenv("JAVA_HOME");
+            String defaultTrustStorePath = javaHome + "/jre/lib/security/cacerts";
+            String destinationTrustStorePath = "/service/var/cacerts";
```
[the remark this answers is not in the document]
Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: Done.

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 10]
```diff
@@ -153,6 +164,62 @@ public static void main(String[] args)
...
+            String certfile = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt";
```
[the remark this answers is not in the document]
Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: Done

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 11]
```diff
@@ -153,6 +164,62 @@ public static void main(String[] args)
...
+            String alias = "kubernetes-root-ca-cert";
+            char[] password = "changeit".toCharArray();
+
+            //Check if the K8S cert exists. If exists, configure new truststore
+            File f = new File(certfile);
+            if(f.exists() && !f.isDirectory()) {
```
[the remark this answers is not in the document]
Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: Loaded code style

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 12]
```diff
@@ -153,6 +164,62 @@ public static void main(String[] args)
...
+                System.out.println("Detected Kubernetes Cert.");
+
+                File source = new File(defaultTrustStorePath);
+                File dest = new File(destinationTrustStorePath);
+                Files.copy(source.toPath(), dest.toPath(), StandardCopyOption.REPLACE_EXISTING);
```
[the remark this answers is not in the document]
Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: I need to know the path of the default store as its relative to java_home. I however, have avoided the copy function.

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 13]
```diff
@@ -153,6 +164,62 @@ public static void main(String[] args)
...
+                FileInputStream is = new FileInputStream(destinationTrustStorePath);
+                KeyStore keystore = KeyStore.getInstance(KeyStore.getDefaultType());
+                keystore.load(is, password);
+                is.close();
+
+                //Get cert
+                CertificateFactory cf = CertificateFactory.getInstance("X.509");
+                InputStream certstream = getCertStream (certfile);
+                Certificate certs =  cf.generateCertificate(certstream);
+                certstream.close();
```
[the remark this answers is not in the document]
Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: Done.

File: launcher/src/main/java/com/proofpoint/launcher/Main.java (hunk 1 above)
Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: Done

File: launcher/src/main/java/com/proofpoint/launcher/Main.java (hunk 3 above)
Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: Done

File: launcher/src/main/java/com/proofpoint/launcher/Main.java (hunk 4 above)
Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: Done

File: launcher/src/main/java/com/proofpoint/launcher/Main.java (hunk 5 above)
Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: Fixed.

File: CHANGES [hunk 14]
```diff
@@ -2106,24 +2104,24 @@ Platform 0.78
 
 * Rack Packaging
 
- - Removed generation of gemfile.jar in favour of using Bundler’s standalone. 
- - Changed tar generation to include specific files instead of by exclusion to prevent pulling in unnecessary 
+ - Removed generation of gemfile.jar in favour of using Bundler’s standalone.
+ - Changed tar generation to include specific files instead of by exclusion to prevent pulling in unnecessary
```
Comment (johngmyers) [inline CHANGES]: Looks like the editor made a bunch of unnecessary changes.

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 15]
```diff
@@ -164,62 +162,56 @@ public static void main(String[] args)
...
-
-                // Save the new truststore contents
-                FileOutputStream out = new FileOutputStream(destinationTrustStorePath);
-                keystore.store(out, password);
-                out.close();
-                System.out.println("Created new TrustStore with Kubernetes Certificate");
-
-                return destinationTrustStorePath;
+        // Copies default truststore to accessable path. Adds k8s ca root cert to truststore. Returns path of new truststore
+        private String addKubernetesToTrustStore(){
```
Comment (johngmyers) [inline launcher/src/main/java/com/proofpoint/launcher/Main.java]: Missing space before {.

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 16]
```diff
@@ -164,62 +162,56 @@ public static void main(String[] args)
...
+            File certFile = new File("/var/run/secrets/kubernetes.io/serviceaccount/ca.crt");
+
+            // Check if the K8S cert exists. If exists, configure new truststore
+            if (certFile.exists() && !certFile.isDirectory()) {
```
Comment (johngmyers) [inline launcher/src/main/java/com/proofpoint/launcher/Main.java]: isFile() ?

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 17]
```diff
@@ -164,62 +162,56 @@ public static void main(String[] args)
...
+
+                // It is not necessarily possible to find the path of current truststore used.
+                // Will request user to set java home and access default truststore path
+                String javaHome = System.getenv("JAVA_HOME");
+                if (javaHome.equals("null")){
```
Comment (johngmyers) [inline launcher/src/main/java/com/proofpoint/launcher/Main.java]: Not the right way to do a null check. `javaHome != null`

Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: I thought it was returning a string 'null'. Fixed.

File: launcher/src/main/java/com/proofpoint/launcher/Main.java (hunk 16 above)
Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: Fixed.

File: launcher/src/main/java/com/proofpoint/launcher/Main.java (hunk 15 above)
Comment (kousik93) [reply launcher/src/main/java/com/proofpoint/launcher/Main.java]: Done

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 18]
```diff
@@ -164,62 +162,56 @@ public static void main(String[] args)
...
+                    //Get cert and add.
+                    BufferedInputStream bis = new BufferedInputStream(certIs);
+                    CertificateFactory cf = CertificateFactory.getInstance("X.509");
+                    Certificate cert = cf.generateCertificate(bis);
+                    keystore.setCertificateEntry(alias, cert);
+
+                    // Save the new truststore contents
+                    keystore.store(destinationOs, password);
+
+                }catch(KeyStoreException | IOException | CertificateException | NoSuchAlgorithmException e ){
```
Comment (johngmyers) [inline launcher/src/main/java/com/proofpoint/launcher/Main.java]: Still not conforming to code style. Did you reformat this section with the code style loaded?

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 19]
```diff
@@ -164,62 +162,56 @@ public static void main(String[] args)
...
+                    throw new RuntimeException("Error configuring TLS TrustStore: " + e);
```
Comment (johngmyers) [inline launcher/src/main/java/com/proofpoint/launcher/Main.java]: `RuntimeException("Configuring TLS TrustStore", e)`

File: launcher/src/main/java/com/proofpoint/launcher/Main.java [hunk 20]
```diff
@@ -463,23 +455,12 @@ else if (line.matches("(?i)\\s*#\\s*allow\\s+spaces\\s*")) {
...
-                Boolean addTrustStoreConfig = true;
-                //If the trustStore arg is passed from config file, ignore this new truststore.
-                for(String s : launcherArgs){
-                    if(s.contains("-Djavax.net.ssl.trustStore=")){
-                        addTrustStoreConfig = false;
-                    }
+            //Add the trust store. If the trustStore arg is passed from config file, ignore this new truststore.
+            if(launcherArgs.stream().noneMatch(s -> s.startsWith("-Djavax.net.ssl.trustStore="))){
+                String trustStorePath = addKubernetesToTrustStore();
+                if (trustStorePath.length() > 0){
```
Comment (johngmyers) [inline launcher/src/main/java/com/proofpoint/launcher/Main.java]: nit: `if (!trustStorePath.isEmpty()) {`

**Comment by johngmyers** [comment]
Squashed and merged
